pinns/src/pinns.c

```diff
@@ -268,20 +295,48 @@ static bool is_host_ns(const char* const optarg) {
   return optarg && !strcmp (optarg, HOSTNS);
 }
 
+/* Mount namespaces can only be bound into unsharable mount namespaces (to
+ * avoid infinite loops), so force the pin_path to be a bind-mount to
+ * itself and then marked as unsharable. */
+static int setup_unbindable_bindpath(const char *pin_path, const char *ns_name) {
+  char bind_root[PATH_MAX];
+
+  // first, verify the /$PATH/$NSns directory exists
+  if (create_bind_root(bind_root, PATH_MAX - 1, pin_path, ns_name) < 0) {
+    return -1;
+  }
+
+  // TODO: Check if bind_root is already a mountpoint
+  // For now, just blindly try to bindmount itself to itself, ignoring
+  // failures.  If this succeeds, we know it's a mountpoint.  If it fails it
+  // probably failed because it's already a mountpoint, and if it's not, the
+  // call to set MS_UNBINDABLE will fail next.
```

Comment:
could we just attempt an `umount` before the `mount` (or even better run in a loop until it succeeds)?  Just to make sure we don't pile up mounts if it runs multiple times.

Reply:
I think doing an `umount` and then mount again would be problematic if `bind_root` is already a mountpoint we don't manage, or if we're reusing the same bind_root to add an additional mount namespace pin to it.  The behavior right now is that if `bind_root` is already mounted (including if it's already a bind-mount to itself like we intend to do the 1st time around), this mount call fails and there's nothing that piles up.

I was looking around for a nice way to check if it's already mounted, but it looked like the only way is to parse and walk through /proc/self/mount[s|info], or to use something like libmount that does this already, and either way it seemed heavy-weight when a try-and-fail-if-it's already-mounted does the job okay.

Reply:
> I was looking around for a nice way to check if it's already mounted, but it looked like the only way is to parse and walk through /proc/self/mount[s|info], or to use something like libmount that does this already, and either way it seemed heavy-weight when a try-and-fail-if-it's already-mounted does the job okay.

agreed.  It is not worth to add that complexity